lib/python/rosie/ws.py

```diff
@@ -19,39 +19,55 @@
 # -----------------------------------------------------------------------------
 """Rosie discovery service.
 
-Classes:
-    RosieDiscoServiceRoot - discovery service root web page.
-    RosieDiscoService - discovery service for a given prefix.
+Base classes:
+    RosieDiscoServiceApplication - collection of request handlers defining the
+        discovery service web application.
+    RosieDiscoServiceRoot - discovery service root web page request handler.
+    RosieDiscoService - discovery service request handler for a given prefix.
+
+Sub-classes, for handling API points by inheriting from RosieDiscoService:
+    HelloHandler - overrides HTTP GET method to write a hello message
+    SearchHandler - overrides HTTP GET method to serve a database search
+    QueryHandler - overrides HTTP GET method to serve a database query
 
 """
 
-import cherrypy
-from isodatetime.data import get_timepoint_from_seconds_since_unix_epoch
 import jinja2
 import json
-from rose.host_select import HostSelector
+import socket
+import tornado.httpserver
+import tornado.ioloop
+from tornado.options import define, options, parse_command_line
+import tornado.web
+
+from isodatetime.data import get_timepoint_from_seconds_since_unix_epoch
 from rose.resource import ResourceLocator
 import rosie.db
 from rosie.suite_id import SuiteId
 
 
-class RosieDiscoServiceRoot(object):
+DEFAULT_PORT = 8080
 
-    """Serves the Rosie discovery service index page."""
+
+class RosieDiscoServiceApplication(tornado.web.Application):
+
+    """Basic Tornado application defining the web service."""
 
     NS = "rosie"
     UTIL = "disco"
     TITLE = "Rosie Suites Discovery"
 
     def __init__(self, *args, **kwargs):
-        self.exposed = True
+        self._ioloop = None
+        self.httpserver = None
+
         self.props = {}
         rose_conf = ResourceLocator.default().get_conf()
         self.props["title"] = rose_conf.get_value(
             ["rosie-disco", "title"], self.TITLE)
         self.props["host_name"] = rose_conf.get_value(["rosie-disco", "host"])
         if self.props["host_name"] is None:
-            self.props["host_name"] = HostSelector().get_local_host()
+            self.props["host_name"] = socket.gethostname()
             if self.props["host_name"] and "." in self.props["host_name"]:
```

Comment:
Out of curiosity, in what circumstances does `socket.gethostname()` return something with a `.` in it?

Reply:
From the [method description in the docs](https://docs.python.org/3/library/socket.html#socket.gethostname), it infers that a ``.`` is there only in the case of a FQDN (fully qualified domain name), which consists of a host name *and*, appended after a ``.``, the name of a domain where that host is located. There is some good & concise (& seemingly reliable) background including regarding the need cases for a domain component [here](https://kb.iu.edu/d/aiuv). So in the conditional in the subsequent lines:

https://github.com/metomi/rose/blob/9058d426113ce1d2d667964476eab742ab6e65bb/lib/python/rosie/ws.py#L70-L73

we check whether there is an attached domain & if so extract only the host name component.

Reply:
Can use `HostSelector().get_local_host()` again, now that it has completed 2to3.